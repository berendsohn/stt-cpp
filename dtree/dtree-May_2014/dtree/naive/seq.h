// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_SEQ_H_
#define DTREE_NAIVE_SEQ_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndSeq : public Base {
 public:
  EndSeq() : solid_() {}
  EndSeq* solid(int i) const { return solid_[i]; }
  void set_solid(int i, EndSeq* solid_i) { solid_[i] = solid_i; }
  template<typename Forwarder>
  void CopyFrom(const EndSeq& node, Forwarder forward);
 private:
  EndSeq* solid_[2];
  DISALLOW_COPY_AND_ASSIGN(EndSeq);
};

template<typename Group, typename Base>
class WithValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToSeq(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromSeq(Node* node, typename Group::Type delta);

 protected:
  WithValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  template<typename Node>
  static typename Semigroup::Type AggrSeq(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostSeq(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeftmostSeq(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRightmostSeq(Node* node, const Predicate& predicate);
 protected:
  WithAggr() {}
};

template<typename Base>
class WithReverseBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void ReverseSeq(Node* node);
 protected:
  WithReverseBy() {}
};

template<typename Base>
class WithReverse : public Base {
 public:
  template<typename Node>
  static void ReverseSeq(Node* node);
 protected:
  WithReverse() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_SEQ_H_
