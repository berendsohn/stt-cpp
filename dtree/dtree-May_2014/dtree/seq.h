// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_SEQ_H_
#define DTREE_SEQ_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/seq.h"
#else
#include "dtree/selfadjust/seq.h"
#endif
#endif  // DTREE_SEQ_H_
