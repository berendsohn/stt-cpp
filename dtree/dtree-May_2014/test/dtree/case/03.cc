#define DTREE_CONFIG EndSeq<WithReverse<WithAggr<dtree::Max<double>, dtree::Greater_<double>, WithAggr<dtree::Min<double>, dtree::LessEqual_<double>, WithValue<dtree::Add<double>, WithAggr<dtree::CountAndSum<int, int>, dtree::IndexByCount_<int>, WithValue<dtree::Nop<int>, WithAggr<dtree::Count<int>, dtree::Index_<int>, WithAggr<dtree::Count<int>, dtree::Index_<int>, Begin > > > > > > > > >
#include "test/dtree/seq.cc"
