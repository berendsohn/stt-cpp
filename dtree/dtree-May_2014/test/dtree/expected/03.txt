<<<
  1: v1 v9
  9: v0 v-3

  6: v6 v-14

  7: v0 v12
  15: v7 v21
  12: v1 v11
  4: v7 v-1

  10: v5 v-6
  8: v3 v-15
  2: v7 v-6
  11: v1 v6
  3: v2 v-4

  13: v7 v-22
  5: v7 v-5
  14: v6 v-7
>>>
<<<
  @1: a1 a1 v1 aC1S1 v9 a0 a0 f0
    \9: a1 a1 v0 aC1S0 v-12 a0 a0 f1

    /7: a3 a3 v0 aC3S8 v13 a-1 a9 f0
        /12: a1 a1 v1 aC1S1 v-10 a0 a0 f1
      \15: a2 a2 v7 aC2S8 v9 a-10 a0 f1
  @4: a3 a3 v7 aC3S15 v-1 a0 a22 f0

  @6: a1 a1 v6 aC1S6 v-14 a0 a0 f0

    /10: a1 a1 v5 aC1S5 v9 a0 a0 f0
  @8: a4 a4 v3 aC4S13 v-15 a0 a21 f0
      /2: a1 a1 v7 aC1S7 v-12 a0 a0 f0
    \11: a3 a3 v1 aC3S10 v21 a-12 a0 f0
      \3: a1 a1 v2 aC1S2 v-10 a0 a0 f0

  @14: a3 a3 v6 aC3S20 v-7 a-15 a2 f1
    \5: a2 a2 v7 aC2S14 v2 a-17 a0 f0
      \13: a1 a1 v7 aC1S7 v-17 a0 a0 f1
>>>
